snuba/datasets/discover.py

```diff
@@ -299,6 +299,8 @@ def column_expr(
                 column_name = "tags[sentry:dist]"
             if column_name == "user":
                 column_name = "tags[sentry:user]"
+            if column_name == "duration":
+                return "0"
```

Comment:
I would not go down this direction to fix that issue. It seems to me the issue lies in the fact that the boundaries of which queries are valid and which ones are not is not properly defined. If the user asks for duration on an error, that should be an invalid query.

This is a violation of the Snuba data model, it would silently yield wrong results if you ask for duration on any query that somehow gets resolved to the events table since there is no duration there (even if the table contains transactions). Now instead of telling the user their query is wrong we will return a wrong results with the expectation to be correct. Same for any query that ends up in the merged table.

What's the downside of not accepting this sort of queries?

Reply:
We're trying to hide the fact that our data is split between two tables from the user. That means that if a user queries for data that needs to be hydrated from both tables (duration and group_id for example) then we need to return a sane result for the data that will be missing.

Making this an invalid query is super confusing for the user because then they need to know details about our table structure and how we pick between the tables.

Reply:
I would counter the argument that returning 0 for durations when you load data from the errors/events table is a sane answer or any better than an error. Imagine this case:

`select p95(duration), from  ... where p95(duration) > 10  (+ same conditions as the query that highlighted the issue`

If the query goes to the event table, we would return transactions and errors (if we are interested in transactions only we should have gone to the transactions table to begin with). The issue is that transactions will come with duration=0 because there is no duration in the events table since it is not a common column. Now your query won't fail but the result will be silently wrong. If I was a customer I would be a lot more pissed off at this than at a properly explained error.

Taking a step back on this problem. I see what you mean when you say that we are trying to hide the difference between the two tables. But I think more variables to consider that are changing the way this problem is framed:
- subscriptions. They simply cannot work on the single table assumption (because of the consistency guarantees that do not work across tables) and this is prompting a redefinition of the data model
- this is one of those use cases marked as (not supported, to be defined) from the data model definition done in october. Supporting this properly and consistently needs a different storage structure than we have and a different snuba data model. So if now this is becoming an important matter I would urge to clarify those points.

So I would advise against going this way, but, ultimately, this is a product decision. If, after considering the pros-cons above you want to go ahead, I will not block this, but please add a comment at least explaining that we know this will generate wrong results and we accept the consequences.